neonvm/controllers/virtualmachine_controller.go

```diff
@@ -744,55 +747,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
+		if err == nil {
+			// delete current runner
+			if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
 				return err
 			}
-			// do cleanup
+		} else if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		// We must keep the VM status the same until we know the neonvm-runner container has been
+		// terminated, otherwise we could end up starting a new runner pod while the VM in the old
+		// one is still running.
+		//
+		// Note that this is required because 'VmSucceeded' and 'VmFailed' are true if *at least
+		// one* container inside the runner pod has finished; the other may still be running.
+		if apierrors.IsNotFound(err) || runnerContainerStopped(vmRunner) {
+			// clean up status, but keep the phase; we'll use it below (and possibly keep it around to
+			// be visible to the user if we don't intend to restart).
+			phase := virtualmachine.Status.Phase
 			virtualmachine.Cleanup()
-		case vmv1.RestartPolicyOnFailure:
-			log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
-			// get runner to delete
-			found := true
-			vmRunner := &corev1.Pod{}
-			err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
-			if apierrors.IsNotFound(err) {
-				found = false
-			} else if err != nil {
-				return err
+			virtualmachine.Status.Phase = phase
+
+			shouldRestart := virtualmachine.Spec.RestartPolicy == vmv1.RestartPolicyAlways ||
+				(virtualmachine.Spec.RestartPolicy == vmv1.RestartPolicyOnFailure && virtualmachine.Status.Phase == vmv1.VmFailed)
+
+			if shouldRestart {
+				log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
+				virtualmachine.Status.Phase = "" // reset to trigger restart
 			}
-			// delete runner only when VM failed
-			if found && virtualmachine.Status.Phase == vmv1.VmFailed {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
+
+			// TODO for RestartPolicyNever: implement TTL or do nothing
+		}
+	default:
+		// do nothing
+	}
+
+	return nil
+}
+
+type runnerStatus string
+
+const (
+	runnerUnknown   runnerStatus = "Unknown"
+	runnerPending   runnerStatus = "Pending"
+	runnerRunning   runnerStatus = "Running"
+	runnerFailed    runnerStatus = "Failed"
+	runnerSucceeded runnerStatus = "Succeeded"
+)
+
+func runnerContainerStatus(pod *corev1.Pod) runnerStatus {
+	switch pod.Status.Phase {
+	case "", corev1.PodPending:
+		return runnerPending
+	case corev1.PodSucceeded:
+		return runnerSucceeded
+	case corev1.PodFailed:
+		return runnerFailed
+	case corev1.PodUnknown:
+		return runnerUnknown
+
+	// With runners with neonvm-container-mgr, the pod phase itself won't be updated if just the
```

Comment:
What happens exactly if one the containers fail in the VM?

Reply:
If one of the containers fails (but not the other), the the pod will still be `Running`, but the VM _should_ be `Failing`.

The version of this function in this PR is now outdated; #749 has the the updated version, and should be merged before this.